caledonia/src/bounded.rs

```diff
@@ -0,0 +1,451 @@
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: [u8; 32]) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = Vec::new();
+        data.push(r.h.clone());
+        data.push(s.to_vec());
+        let (h, h_usize) = Round::h1(data, r.n_p);
+        Round {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h: h.to_vec(),
+            h_usize,
+            n_p: r.n_p,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    r: usize,
+    /// Proof 2nd counter
+    d: usize,
+    /// Proof tuple
+    items: Vec<[u8; 32]>,
+}
+
+impl Proof {
+    /// Returns an empty proof
+    fn empty() -> Self {
+        Proof {
+            r: 0,
+            d: 0,
+            items: Vec::new(),
+        }
+    }
+
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    // TODO: We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h0(setup: &Setup, v: usize, s: [u8; 32]) -> usize {
+        let mut data = Vec::new();
+        data.push(v.to_ne_bytes().to_vec());
+        data.push(s.to_vec());
+        let digest = utils::combine_hashes(data);
+        // return (digest, utils::oracle(&digest, setup.n_p));
+        return utils::oracle(&digest, setup.n_p);
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut data = Vec::new();
+        data.push(r.v.to_ne_bytes().to_vec());
+        data.push(r.t.to_ne_bytes().to_vec());
+        for s in &r.s_list {
+            data.push(s.clone().to_vec());
+        }
+        let digest = utils::combine_hashes(data);
+        return utils::oracle(&digest, setup.q) == 0;
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &Vec<Vec<[u8; 32]>>,
+        round: &Round,
+        limit: &mut usize,
+    ) -> Option<Proof> {
+        // If the round candidate is at the last iteration (len(round) == u)
+        if round.s_list.len() == setup.u {
+            // if H2(t, x_0, ..., x_u) = true, return candidate as proof
+            if Proof::h2(setup, round) {
+                let r = round.v;
+                let d = round.t;
+                let items = round.s_list.clone();
+                return Some(Proof { r, d, items });
+            } else {
+                return None;
+            }
+        }
+        // Otherwise, update round candidate with all s in bins[H1(t, x_1, ..., x_i)] and continue
```

Comment:
I would remove the comment here too. `Otherwise` seems to imply we do this _only_ if we don't do the other thing, which is not th case.